### ISSUE TITLE
Averaging mask in Contextual Attention layer

### ISSUE THREAD

**Issue by arthurhero** [JiahuiYu/generative_inpainting#233]
Hey Jiayu,

Here is another question! (Thank you for your speedy response this afternoon lol)

In inpaint_ops.py line 271, it says `mm = tf.cast(tf.equal(tf.reduce_mean(m, axis=[0,1,2], keep_dims=True), 0.), tf.float32)`

The 0th axis should be the batch size, I remember. I understand why we are averaging through the kernel, but why are we averaging along the batch as well? Don't different images have different masks in a batch?

Thanks!

**Comment by arthurhero** [comment]
Ah, wait... is it because you are always using a same mask for all images in one batch?

**Comment by JiahuiYu** [comment]
Yes in this version, the masks are the same in each mini batch. It would be great if you can modify to support multi-mask. I guess that may lead to better performance.

**Comment by arthurhero** [comment]
> Yes in this version, the masks are the same in each mini batch. It would be great if you can modify to support multi-mask. I guess that may lead to better performance.

Finished implementing it in Pytorch! Will release the code... once the model converges lol

**Comment by FloCF** [comment]
@arthurhero : I am also working on a Pytorch implementation. Everything is more or less straight forward - thanks to the many question @JiahuiYu answered - except for the centex-atten layer (probably because of my limited tensorflow skills). Also for my implementation I need flexible masks within a batch. I would therefore be really grateful if you could share your context-atten pytorch implementation.

**Comment by arthurhero** [comment]
> @arthurhero : I am also working on a Pytorch implementation. Everything is more or less straight forward - thanks to the many question @JiahuiYu answered - except for the centex-atten layer (probably because of my limited tensorflow skills). Also for my implementation I need flexible masks within a batch. I would therefore be really grateful if you could share your context-atten pytorch implementation.

Heyy. I've finished this layer long time ago but forgot to make it public lol. Will give you link very soon.

**Comment by arthurhero** [comment]
> @arthurhero : I am also working on a Pytorch implementation. Everything is more or less straight forward - thanks to the many question @JiahuiYu answered - except for the centex-atten layer (probably because of my limited tensorflow skills). Also for my implementation I need flexible masks within a batch. I would therefore be really grateful if you could share your context-atten pytorch implementation.

https://github.com/arthurhero/deep_fill_2_pytorch

**Comment by JiahuiYu** [comment]
@arthurhero Thanks for your help!
